ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -147,30 +144,27 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
 
     @Override
     public WithFailures.Result<PDone, PubsubMessage> expand(PCollection<PubsubMessage> input) {
-      ValueProvider<DynamicPathTemplate> pathTemplate = NestedValueProvider.of(outputPrefix,
-          DynamicPathTemplate::new);
-      ValueProvider<String> staticPrefix = NestedValueProvider.of(pathTemplate,
-          value -> value.staticPrefix);
+      DynamicPathTemplate pathTemplate = new DynamicPathTemplate(outputPrefix);
+      String staticPrefix = pathTemplate.staticPrefix;
 
       FileIO.Write<List<String>, PubsubMessage> write = FileIO
           .<List<String>, PubsubMessage>writeDynamic()
           // We can't pass the attribute map to by() directly since MapCoder isn't
           // deterministic;
           // instead, we extract an ordered list of the needed placeholder values.
           // That list is later available to withNaming() to determine output location.
-          .by(message -> pathTemplate.get()
+          .by(message -> pathTemplate
               .extractValuesFrom(DerivedAttributesMap.of(message.getAttributeMap())))
           .withDestinationCoder(ListCoder.of(StringUtf8Coder.of())) //
           .withCompression(compression) //
           .via(Contextful.fn(format::encodeSingleMessage), TextIO.sink()) //
           .to(staticPrefix) //
-          .withNaming(placeholderValues -> NoColonFileNaming.defaultNaming(
-              pathTemplate.get().replaceDynamicPart(placeholderValues), format.suffix()));
+          .withNaming(placeholderValues -> NoColonFileNaming
+              .defaultNaming(pathTemplate.replaceDynamicPart(placeholderValues), format.suffix()));
 
       if (inputType == InputType.pubsub) {
-        // Passing a ValueProvider to withNumShards disables runner-determined sharding, so we
-        // need to be careful to pass this only for streaming input (where runner-determined
-        // sharding is not an option).
+        // withNumShards disables runner-determined sharding, so we need to be careful to pass this
+        // only for streaming input (where runner-determined sharding is not an option).
         write = write.withNumShards(numShards);
       }
```

Comment:
I'm unclear on whether putting this in a condition is still necessary, because we aren't using ValueProvider, but we do set a default of `100`.

Reply:
The docs for `withNumShards(int numShards)` specify:

> 0 means runner-determined sharding. Specifying a non-zero value may hurt performance, because it will limit the parallelism of writing

So I think the right thing to do now is to change our default to 0 and document that 0 means runner-determined sharding.

Reply:
Actually, I take that back. We should still ignore `numShards` in the batch case. This conditional still makes sense, and it still makes sense to have 100 as the default value that we pass when we have a streaming input.